# Einsendeaufgabe 1: Numerische Genauigkeit und Gleichungssysteme (100 Punkte)
In dieser Aufgabe sollen Sie ein wenig mehr Erfahrung mit NumPy und numerischen Methoden gewinnen.  
Zur Erinnerung empfehle ich an dieser Stelle, die Definition der [IEEE-Fließkommazahlen](https://de.wikipedia.org/wiki/IEEE_754) zu wiederholen.  

## Addition von Zahlen (20 Punkte) 

Gegeben sei ein Array *array*, dass 100 mal die Zahl $10^{-16}$ enthält und einmal (als ersten Eintrag) die Zahl $1$. 

In [2]:
import numpy as np

array = np.concatenate(([1], np.full(100, 1e-16)))
print(array)

[1.e+00 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16 1.e-16
 1.e-16]


**Aufgabe:** Addieren Sie alle Werte in `array`:
- mit der Funktion `np.sum()`
- mit einer Schleife, die von **vorne nach hinten** über `array` iteriert
- mit einer Schleife, die von **hinten nach vorne** über `array` iteriert

_Points:_ 6

## Addition von Zahlen im Array: Unterschiedliche Berechnungsmethoden

In dieser Aufgabe geht es um die Addition von Zahlen, die in einem Array gespeichert sind. Das Array besteht aus 101 Einträgen, wobei der erste Eintrag die Zahl 1 enthält und die restlichen 100 Einträge jeweils den Wert $ 1 \times 10^{-16} $ haben. Die Aufgabe besteht darin, die Summe der Werte in diesem Array auf verschiedene Arten zu berechnen: einmal mit der `np.sum()`-Funktion und zweimal unter Verwendung von Schleifen, die das Array jeweils von vorne nach hinten bzw. von hinten nach vorne durchlaufen.

In [3]:
sum_numpy = np.sum(array)
print(f"{sum_numpy=}")

sum_forward = 0.0
for x in array:
    sum_forward += x
print(f"{sum_forward=}")

sum_reverse = 0.0
for x in reversed(array):
    sum_reverse += x
print(f"{sum_reverse=}")

sum_numpy=np.float64(1.0000000000000084)
sum_forward=np.float64(1.0)
sum_reverse=np.float64(1.00000000000001)


<!-- BEGIN QUESTION -->

**Aufgabe:** Erklären Sie die Ergebnisse. Wie werden die Zahlen 1 und 1e-17 im Computer dargestellt? (ausführliche Erläuterung erforderlich!)  

_Points:_ 12

## Numerische Genauigkeit von verschiedenen Additionsverfahren

### Unterschiede in der Summierung

Bei der Berechnung der Summe mit unterschiedlichen Verfahren ergeben sich leicht unterschiedliche Ergebnisse:

```python
sum_numpy   = np.float64(1.0000000000000084)
sum_forward = np.float64(1.0)
sum_reverse = np.float64(1.00000000000001)
```
Diese Unterschiede entstehen durch **Rundungsfehler** in der **Gleitkommadarstellung**, die auf der begrenzten Anzahl an Bits basiert, mit der reale Zahlen intern im Computer dargestellt werden.

---

## Gleitkommazahlen im IEEE 754 Standard

Python und NumPy verwenden standardmäßig `float64` (IEEE 754 Double Precision), das wie folgt aufgebaut ist:

- **1 Bit** für das Vorzeichen  
- **11 Bits** für den Exponenten  
- **52 Bits** für die Mantisse  

Insgesamt also **64 Bit**. Die reale Zahl wird durch folgende Formel dargestellt:

```
Zahl = (-1)^s · 1.m · 2^(e - 1023)
```

Dabei ist:

- `s` das Vorzeichenbit  
- `m` die Mantisse (ohne führende 1)  
- `e` der gespeicherte Exponent mit **Bias** = 1023  

---

## Maschinengenauigkeit

Die kleinste Zahl, die zu `1.0` addiert werden kann, sodass sich das Ergebnis **numerisch unterscheidet**, nennt man **Maschinengenauigkeit** oder **machine epsilon** $\varepsilon_{\text{mach}}$:

```python
np.finfo(float).eps
# Ergebnis: 2.220446049250313e-16
```

Zahlen kleiner als `eps` „verpuffen“ bei der Addition zu größeren Zahlen wie `1.0`, da sie außerhalb der darstellbaren Genauigkeit der Mantisse liegen. Wenn zwei Gleitkommazahlen stark unterschiedlich groß sind, kann die kleinere Zahl beim Addieren „verschluckt“ werden – das nennt man Verlust der Signifikanz:

```python
print(1e-16 < np.finfo(float).eps)  # True
print(1.0 + 1e-16 == 1.0)           # True
```

---

## Warum liefert `sum_forward` ein anderes Ergebnis als `sum_reverse`?

Angenommen:

```python
array = np.array([1.0] + [1e-16] * 100)
```

- `sum_forward` addiert zuerst `1.0` und dann viele kleine `1e-16`. Diese einzelnen Summanden sind kleiner als `eps` und **gehen verloren**.
- `sum_reverse` addiert zuerst die kleinen `1e-16` zu einem größeren Zwischenwert (z. B. `1e-14`), der dann zu `1.0` addiert wird und sich **merklich auswirkt**.
- `sum_numpy` verwendet oft optimierte oder paarweise Additionen (`pairwise summation`), was ebenfalls zu einem etwas besseren Ergebnis führen kann.

Gleitkommazahlen sind **nicht assoziativ**, d. h.:

```
(a + b) + c ≠ a + (b + c)
```

Das bedeutet, dass die Reihenfolge der Addition das Ergebnis beeinflussen kann, besonders wenn sehr kleine Zahlen in der Summe enthalten sind.

---

## Darstellung der Zahl `1.0` im IEEE 754-Standard

#### Mathematische Form:

$
1.0 = 1.0 \cdot 2^0 \\
\Rightarrow \text{Echter Exponent} = 0 \\
\Rightarrow \text{Gespeicherter Exponent} = 0 + \text{Bias} = 1023 = 01111111111_2
$

#### Mantisse:
- Die Zahl `1.0` entspricht im Binärsystem: $ 1.000\ldots0_2 $
- Die **führende 1 vor dem Komma** wird im IEEE 754-Format **nicht gespeichert** (implizit angenommen)
- Alle 52 Bits der Mantisse (also der Nachkommastellen) sind daher **0**

#### IEEE 754-Darstellung (64 Bit – `double precision`):

|      s      |   e (11 Bit)   |                    m (52 Bit)                          |
|-------------|----------------|--------------------------------------------------------|
|     `0`     |  `01111111111` | `0000000000000000000000000000000000000000000000000000` |

#### Zusammenfassung:
- **Vorzeichenbit (`s`)**: `0` → Zahl ist positiv  
- **Exponent (`e`)**: 1023 → gespeichert als `01111111111`  
- **Mantisse (`m`)**: nur Nullen (weil keine Nachkommastellen vorhanden sind)  
- Daraus ergibt sich der dargestellte Wert:
  
$
(-1)^0 \cdot 1.0 \cdot 2^{0} = 1.0
$

---

## Darstellung der Zahl `1e-17` im IEEE 754-Standard (Double Precision)

### Mathematische Zerlegung von `1e-17`

### Ziel: Darstellung in der Form

Wir suchen $m$ und $e$, sodass:

$1e-17 = 1.m \cdot 2^{e - 1023}$

### Schritt 1: Umwandlung in Zweierpotenz

Zunächst schreiben wir `1e-17` in etwa als Zweierpotenz:

$1e-17 = 10^{-17} \approx (2^{3.321928094887354...})^{-17}\approx 2^{-3.321928094887354... \cdot 17} \approx 2^{-56.47}$

Das bedeutet:

$1e-17 \in (2^{-57}, 2^{-56})$

### Schritt 2: Rundung auf nächstliegende Zweierpotenz

Da `1e-17` bzw. $2^{-56.47}$ näher an $2^{-57}$ liegt, weil:

$\lvert 2^{-56.47}-2^{-57} \rvert = \lvert  \frac{1}{2^{56.47}} - \frac{1}{2^{57}} \rvert < \lvert  \frac{1}{2^{56.47}} - \frac{1}{2^{56}} \rvert= \lvert 2^{-56.47} - 2^{-56} \rvert$

gilt:

$1e-17 \approx 1.m \cdot 2^{-57}$

### Schritt 3: Exponent berechnen

Der effektive Exponent ist:

$e_{\text{effektiv}} = -57$

Im IEEE 754 Double Precision wird der **Exponent mit Bias 1023** gespeichert:

$e_{\text{gespeichert}} = -57 + 1023 = 966$

In Binär:

$966_{10} = 01111000110_2$

### Schritt 4: Berechnung der Mantisse

Es gilt $1e-17 \approx 1.m \cdot 2^{-57}$ und somit $1.m = \frac{1e-17}{2^{-57}} = 1.4411518807585587...$

Wir nehmen die Nachkommastellen von `1.44115188...` und wandeln sie in Binär um. 

Wir schreiben `1e-17` in binärer Normalform:

$1e-17 = 1.0111000011101111010101000110010001101101010010010110\dots_2 \cdot 2^{-57}$

Die Mantisse sind **die 52 Bits nach dem Komma** (die führende $1.$ wird nicht gespeichert):

### Schritt 5: Zusammensetzung der IEEE-754-Bitfolge

| Feld        | Wert                                                         |
|-------------|--------------------------------------------------------------|
| Vorzeichen  | `0` (da positiv)                                             |
| Exponent    | `01111000110` (`11` Bit für `966`)                           |
| Mantisse    | `0111000011101111010101000110010001101101010010010110`       |


### Gesamter 64-Bit-IEEE-754-Code (Binär)

Die IEEE-754-Bitfolge sieht wie folgt aus (in 64 Bit):
|      s      |   e (11 Bit)   |                    m (52 Bit)                          |
|-------------|----------------|--------------------------------------------------------|
|     `0`     |  `01111000110` | `0111000011101111010101000110010001101101010010010110` |

---
### Fazit:

- Die Zahl `1.0` kann exakt im Format IEE 754 `float64` dargestellt werden.  
- Die Zahl `1e-17` ist **zu klein** und hat eine **unendliche Binärdarstellung**, was zu einer _Rundung_ führen muss. Eine präzise Darstellung innerhalb der 52 Bits der Mantisse ist nicht möglich. 
---

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

**Aufgabe:** Welches Verfahren verwendet `np.sum()` für die Addition?

_Points:_ 2

## Kaskadensummation

Paarweise Summation (auch Kaskadensummation genannt) ist ein numerisch besseres Verfahren, bei dem die Werte in mehreren Schritten summiert werden. Dies bedeutet, dass Werte mit ähnlichen Größen zuerst zusammengefasst werden. Dadurch wird der Fehler reduziert, der entstehen kann, wenn immer kleinere Zahlen zu einem sehr großen Wert hinzugefügt werden (wie es bei der sequentiellen Summation der Fall wäre).

### Verfahren von `np.sum()`

NumPy verwendet für die Addition von Arrays ein Verfahren namens **Kaskadensummation** oder **paarweise Summation**. Dieses Verfahren reduziert Rundungsfehler im Vergleich zur sequentiellen Summation und führt in vielen Fällen zu einer höheren Genauigkeit. `np.sum()` kann interne Optimierungen verwenden, wie z. B. parallelisierte Summation, abhängig von der Größe des Arrays und der Hardware. In modernen Implementierungen von NumPy wird oft die `np.add.reduce()` Funktion verwendet, um die Summation durchzuführen, die im Hintergrund mit dem `ufunc`-Mechanismus arbeitet. Diese `reduce`-Methode summiert die Werte sequenziell, aber auch sie optimiert den Prozess, indem sie die Reihenfolge der Summanden variiert oder bei Bedarf parallelisiert.


Mehr Informationen zu dieser Technik gibt es auf [Wikipedia](https://en.wikipedia.org/wiki/Pairwise_summation) und in der [NumPy-Dokumentation](https://numpy.org/doc/stable/reference/generated/numpy.sum.html).

<!-- END QUESTION -->

## Grenzwerte von Funktionen (20 Punkte)

Geben sei die Funktion $f(x)=\frac{e^x-1}{x}$

Es gilt: $lim_{x \to 0}f(x) = 1$

**Aufgabe:** Schreiben Sie eine Python Funktion `f(x)`, mit deren Hilfe Sie $f(x)$ für $x= 10^{-1}, 10^{-2}, \cdots , 10^{-15}$ mit NumPy berechnen und geben Sie das Ergebnis aus.

**Aufgabe:** Berechnen Sie nun die Funktion für  $x= 10^{-16}, 10^{-17}, \cdots , 10^{-20}$

_Points:_ 8

<h2>Berechnung der Funktion f(x) = (exp(x) - 1) / x </h2>

In dieser Aufgabe soll die Funktion f(x) für x = 10<sup>-1</sup>, 10<sup>-2</sup>, ..., 10<sup>-15</sup> berechnet werden. Danach soll die Funktion für noch kleinere Werte von x, nämlich x = 10<sup>-16</sup>, 10<sup>-17</sup>, ..., 10<sup>-20</sup>, berechnet werden. 

Der Grenzwert der Funktion für x → 0 ist 1. Dabei sollen mögliche numerische Probleme bei sehr kleinen x beobachtet werden.

In [ ]:
import numpy as np

# Definition der Funktion f(x) = (e^x - 1)/x
def f(x):
        return (np.exp(x) - 1) / x

# Bereich 1: x = 10^(-1) bis 10^(-15)
# Erzeuge die Werte x = 10^(-1), 10^(-2), ..., 10^(-15)
input_numbers = np.array([10**(-i) for i in range(1, 16)])
result_limits = f(input_numbers)

#Ausgabe
for x_val, fx in zip(input_numbers, result_limits):
# Die Anwendung von zip gibt einen Iterator zurück, der in der Lage ist Tupel zu erzeugen
    print(f"x = {x_val:.0e}, f(x) = {fx}")

x = 1e-01, f(x) = 1.0517091807564771
x = 1e-02, f(x) = 1.005016708416795
x = 1e-03, f(x) = 1.0005001667083846
x = 1e-04, f(x) = 1.000050001667141
x = 1e-05, f(x) = 1.000005000006965
x = 1e-06, f(x) = 1.0000004999621837
x = 1e-07, f(x) = 1.0000000494336803
x = 1e-08, f(x) = 0.999999993922529
x = 1e-09, f(x) = 1.000000082740371
x = 1e-10, f(x) = 1.000000082740371
x = 1e-11, f(x) = 1.000000082740371
x = 1e-12, f(x) = 1.000088900582341
x = 1e-13, f(x) = 0.9992007221626409
x = 1e-14, f(x) = 0.9992007221626409
x = 1e-15, f(x) = 1.1102230246251565


In [ ]:
# Bereich 2: x = 10^(-16) bis 10^(-20)
input_numbers_2 = np.array([10**(-i) for i in range(16, 21)])
result_limits_2 = f(input_numbers_2)

#Ausgabe
for x_val, fx in zip(input_numbers_2, result_limits_2):
# Die Anwendung von zip gibt einen Iterator zurück, der in der Lage ist Tupel zu erzeugen
    print(f"x = {x_val:.0e}, f(x) = {fx}")

x = 1e-16, f(x) = 0.0
x = 1e-17, f(x) = 0.0
x = 1e-18, f(x) = 0.0
x = 1e-19, f(x) = 0.0
x = 1e-20, f(x) = 0.0


<!-- BEGIN QUESTION -->

**Aufgabe:** Was fällt auf? Welche Erklärung haben Sie für das Ergebnis? (Ausführliche Erklärung erforderlich) 

_Points:_ 12

<h2> Erklärung des Verhaltens der Funktion bei kleinen x-Werten </h2>

Für Werte von x zwischen `1e-01` bis `1e-14` liefert die Funktion erwartungsgemäß Ergebnisse nahe bei **1**. Ab etwa 1e-15 treten jedoch drastische Änderungen auf, die durch Rundungsfehler bei der Berechnung von `e^x - 1` bedingt sind. Besonders bei sehr kleinen x-Werten wird der Unterschied zwischen `e^x` und `1` auf dem Computer nicht mehr erkennbar, was zu fehlerhaften Ergebnissen führt. Eine stabilere Berechnung kann durch die Verwendung der Funktion `np.expm1(x)` von NumPy erreicht werden.


Von `1e-01` bis `1e-14` ergibt die Funktion erwartungsgemäß Werte, die sehr nahe bei **1** liegen.  Doch ab etwa `1e-15` ändert sich das Verhalten drastisch:

x = 1e-15, f(x) = 1.1102230246251565   
x = 1e-16, f(x) = 0.0  
x = 1e-17, f(x) = 0.0  
...

**Bei genauerer Betrachtung** sieht man, dass bereits ab `1e-08` praktisch keine Annäherung mehr an den Soll-Grenzwert stattfindet. Die Berechnung "divergiert numerisch".

| Phänomen                    | Beschreibung                                                                |
|-----------------------------|------------------------------------------------------------------------------|
| Katastrophale Auslöschung   | Verlust signifikanter Ziffern durch Subtraktion ähnlicher Zahlen             |
| Numerische Instabilität     | Instabiles Verhalten bei kleinen Änderungen der Eingabe                      |
| Rundungsfehlerakkumulation  | Anhäufung von Maschinenfehlern durch viele Rechenoperationen                 |
| (Numerische) Divergenz      | Umgangssprachlich, wenn das Ergebnis sich vom Sollwert entfernt              |

### Ergebnisbeobachtung

Ab etwa $x \approx 10^{-16}$ treten numerische Rundungsfehler sichtbar auf, abhängig von der Maschinengenauigkeit (`float64`).

<h3> Erklärung: Rundungsfehler </h3>

Für sehr kleine `x` gilt näherungsweise:  
`exp(x) ≈ 1 + x`

Wenn `x` so klein wird, dass `e^x` und `1` auf dem Computer **nicht mehr unterscheidbar** sind, wird der Ausdruck `e^x - 1` **numerisch ungenau** (_null_), da beide Terme sehr nah beieinander liegen → Verlust signifikanter Stellen.

> Beispiel:  
> `x = 1e-17` → `exp(x) ≈ 1.00000000000000001`  
> Die Differenz `e^x - 1` ist kleiner als die **Maschinengenauigkeit** `ε` (Epsilon).  
> Deshalb wird `e^x - 1` vom Computer als **0.0** berechnet, da ein 64-Bit-Gleitkommawert (float64) **nicht mehr ausreichend genau** ist.

<h3> Lösung: Stabilere Berechnung mit NumPy </h3>

NumPy bietet mit `np.expm1(x)` eine spezielle Funktion an, die den Ausdruck `exp(x) - 1` **numerisch stabiler** berechnet – insbesondere für sehr kleine `x`.

In [3]:
def f_stabil(x):
    return np.expm1(x) / x

input_numbers_3 = np.array([10**(-i) for i in range(10, 20)])
result_limits_3 = f_stabil(input_numbers_3)

#Ausgabe
for x_val, fx in zip(input_numbers_3, result_limits_3):
# Die Anwendung von zip gibt einen Iterator zurück, der in der Lage ist Tupel zu erzeugen
    print(f"x = {x_val:.0e}, f(x) = {fx}")

x = 1e-10, f(x) = 1.00000000005
x = 1e-11, f(x) = 1.000000000005
x = 1e-12, f(x) = 1.0000000000005
x = 1e-13, f(x) = 1.00000000000005
x = 1e-14, f(x) = 1.000000000000005
x = 1e-15, f(x) = 1.0000000000000007
x = 1e-16, f(x) = 1.0
x = 1e-17, f(x) = 1.0
x = 1e-18, f(x) = 1.0
x = 1e-19, f(x) = 1.0


## Detaillierte Erklärung der Gleitkommadarstellung und Rundungsfehler

Die Präzision einer Gleitpunktzahl hängt von der Mantisse ab. Diese bildet neben dem Exponenten, der Basis und dem Vorzeichen die Gleitpunktzahl.

Solange man nicht explizit davon abweicht, wird `float64` in Python und NumPy verwendet für Gleitpunktzahlen.

- Eine `float32`-Mantisse besteht aus 23 Bit und eine `float64`-Mantisse aus 52 Bit (bei 64 Bit bleiben noch 11 Bit für den Exponenten und 1 Bit für das Vorzeichen).

Bei Werten, die mehr Bits zur Darstellung benötigen als die Mantisse zur Verfügung stellt, kommt es zu Rundungsfehlern, wie in den folgenden Beispielen erklärt:

## Fazit zur Gleitkommadarstellung und Rundungsfehlern

Die Präzision von Gleitkommazahlen ist durch die Anzahl der Bits in der Mantisse begrenzt, was zu Rundungsfehlern führt, wenn Zahlen mit mehr erforderlichen Bits dargestellt werden müssen. Besonders bei sehr kleinen Werten kann dies zu Verlusten an Genauigkeit führen, die sich durch den Einsatz stabiler Berechnungsmethoden wie z. B. `np.expm1(x)` verhindern lassen.

In [3]:
import numpy as np

def mindest_mantissenbits(x, y):
    # Berechnet die Anzahl der benötigten Mantissenbits für eine korrekte Berechnung
    # float128 um exaktere Praezision zu ermoeglichen
    return np.abs(np.log2(np.float128(x)) - np.log2(np.float128(y)))

# Beispiel zur Darstellbarkeit float32
# Mantisse = 1 wenn dargestellte Zahl Potenz von 2, also darstellbar
x = 1
a = np.float128(2**24-1)
print(x + a, np.float32(x+a), "darstellbar?:", mindest_mantissenbits(x,a) < 23 + 1)
# Ergebnis: (16777216.0, 16777216.0, 'darstellbar?:', True)

# Beispiel zur Darstellbarkeit float32
# Dargestellte Zahl muss hier mit Mantisse von Mindestlaenge 24 dargestellt werden, nicht moeglich bei float32
b = np.float128(2**24)
print(x + b, np.float64(x+b), "darstellbar?:", mindest_mantissenbits(x,b) < 23 + 1)
# Ergebnis: (16777217.0, 16777216.0, 'darstellbar?:', False)

# Diese Beispiele verdeutlichen, wie die Genauigkeit und die Darstellung von Zahlen durch die verwendete Gleitkommadarstellung beeinflusst werden.

# Weitere Beispiele zur Rundungsfehlerdarstellung
# In den folgenden Beispielen wird die Auswirkung von kleinen Zahlen in Gleitkommadarstellungen demonstriert:

# Beispiel 1 + 1e-16, vorkommend bei Summierung in Schleife von vorne
c = np.float128(1e-16)
print(x + c, np.float64(x + c), "darstellbar?:", mindest_mantissenbits(x, c) < 52 + 1)
# Ergebnis: (1.0000000000000001, 1.0, 'darstellbar?:', False)

# Beispiel 1 + 1e-16 * 100, vorkommend bei Summierung in Schleife von hinten
d = np.float128(1e-16 * 100)
print(x + d, np.float64(x + d), "darstellbar?:", mindest_mantissenbits(x, d) < 52 + 1)
# Ergebnis: (1.00000000000001, 1.00000000000001, 'darstellbar?:', True)


16777216.0 16777216.0 darstellbar?: True
16777217.0 16777217.0 darstellbar?: False
1.0000000000000001 1.0 darstellbar?: False
1.00000000000001 1.00000000000001 darstellbar?: True


<!-- END QUESTION -->

## Gauß-Verfahren und LU-Zerlegung (60 Punkte) 
Obwohl man, wie wir sehen werden, lineare Gleichungssysteme einfach mit SciPy lösen kann, wollen wir in dieser Aufgabe  einmal selbst eine sogenannte LU-Zerlegung bzw. das Gauß-Verfahren implementieren. Außerdem werden wir diesen Code auch später noch verwenden, wenn wir uns mit Performance-Optimierung und Parallelisierung beschäftigen.

Das Gauß-Verfahren ist ein wichtiges Verfahren zum Lösen von linearen Gleichungssystemen, wir gehen hier schrittweise mit der Implementierung vor.

**Aufgabe:** Implementieren Sie die Funktion `b_solve` zur Lösung einer Dreiecksgleichung der Form:

$\left[ {\begin{array}{cccc} 
a_{11} & a_{12} & \cdots & a_{1n}\\ 
0 & a_{22} & \cdots & a_{2n}\\ 
\vdots & \vdots & \ddots & \vdots \\ 
0 & 0 & \cdots & a_{nn}\\ 
\end{array} } \right] \left[{\begin{array}{c}  
x_1\\ 
x_2\\ 
\vdots\\ 
x_n\\ 
\end{array}}\right] =  
\left[{\begin{array}{c}  
b_1\\ 
b_2\\ 
\vdots\\ 
b_n\\ 
\end{array}}\right]$ 

Hier beginnt man mit der Löung der letzten Zeile, denn es gilt:

$x_n = b_n/a_{nn} $
und dann

$x_{n-1} =  (b_{n-1}-a_{(n-1),n}*x_{n})/a_{(n-1),(n-1)}$

$x_{n-2} =  (b_{n-2}-(a_{(n-2),n}*x_{n}+a_{(n-2),(n-1)}*x_{n-1})/a_{(n-2),(n-2)}$


*Hinweis*: Achten Sie darauf, dass ihr Code nicht nur korrekt, sondern auch effizient ist. 

_Points:_ 12

<span style="color:blue">

  <h2>Backsubstitution-Verfahren</h2>

Im Folgenden wird eine Funktion zur Lösung eines linearen Gleichungssystems mit einer oberen Dreiecksmatrix implementiert. Dieses sogenannte Backsubstitution-Verfahren stellt einen zentralen Schritt im Rahmen des Gauß-Verfahrens bzw. der LU-Zerlegung dar. Dabei wird die Lösung des Systems zeilenweise von unten nach oben bestimmt. Die Implementierung legt besonderen Wert auf numerische Stabilität und Effizienz. 

Die Funktion `b_solve` implementiert eine effiziente Rückwärtseinsetzung zur Lösung eines linearen Gleichungssystems mit oberer Dreiecksmatrix – kompakt, direkt und ohne überflüssigen Overhead.
Um auch Matrizen lösen zu können, welche nicht in oberer Dreiecksform vorliegen, wird der Parameter upper_triangular auf False gesetzt. Dann wird Matrix A mittels der Funktion make_triangular in eine obere Dreiecksmatrix transformiert.

</span>

In [7]:
import numpy as np

def b_solve(A, b, upper_triangular=True):
    """Löst ein lineares Gleichungssystem A * x = b"""

    x = np.zeros_like(b, dtype=float)  # Vektor für die Lösung, initialisiert mit Nullen
    A = A.astype(float)  # Sicherstellen, dass mit floats gerechnet wird
    b = b.astype(float)
    
    #Matrix in Dreiecksform umwandeln wenn nötig
    if not upper_triangular:
        make_triangular(A, b)        
    
    # Rückwärtseinsetzung
    for i in range(len(b) - 1, -1, -1):  
        # Berechnung der Summe der Produkte der bekannten x-Werte
        summation = np.dot(A[i, i + 1:], x[i + 1:]) # hier wird die Vektorisierung, die numpy zur Verfügung stellt, benutzt, um die Effizienz zu erhöhen
        x[i] = (b[i] - summation) / A[i, i]  # Berechnung des aktuellen x_i
        #print(f"x[{i}] = (b[{i}] - summation) / A[i, i] = ({b[i]} - {np.dot(A[i, i + 1:], x[i + 1:])}) / {A[i,i]} = {x[i]}")
    return x

def make_triangular(A, b):
    """bringt Matrix A in Dreiecksform unter Berücksichtigung von Vektor b"""
    
    # Vorwärtselimination (Dreiecksform erzeugen)
    for i in range(len(b)):
        # Pivotisierung: größtes Element in der Spalte nach unten holen (numerische Stabilität)
        max_row = i + np.argmax(abs(A[i:, i]))
        if A[max_row, i] == 0:
            raise ValueError("Das Gleichungssystem hat keine eindeutige Lösung")
        if max_row != i:
            A[[i, max_row]] = A[[max_row, i]]
            b[[i, max_row]] = b[[max_row, i]]

        # Elimination
        for j in range(i + 1, len(b)):
            factor = A[j, i] / A[i, i]
            A[j, i:] -= factor * A[i, i:]
            b[j] -= factor * b[i]

# Beispiel
A = np.array([[2, 3, 5, 1],
              [0, 4, 6, 2],
              [0, 0, 7, 3],
              [0, 0, 0, 8]], dtype=float)

b = np.array([9, 8, 7, 8], dtype=float)

x = b_solve(A, b)

# Testausgabe
print("Lösung x:", x)
print("Prüfung A @ x:", np.dot(A, x))  # Sollte gleich b sein

Lösung x: [1.60714286 0.64285714 0.57142857 1.        ]
Prüfung A @ x: [9. 8. 7. 8.]


**Aufgabe:** Implementieren Sie die LU-Zerlegung einer Matrix mit Pivotisierung.  
Die Funktion `lu_decomposition` soll die Matrizen L und U berechnen.  
Zu Beginn wird der Funktion eine Matrix A übergeben.  


Implementieren Sie die Funktion "in-place", d.h. die Ergebnisse für L und U werden direkt in A geschrieben.  
Dabei wird die ursprüngliche Matrix überschrieben.  
Die Funktion gibt den Pivot-Vektor zurück, der die folgende Form haben sollte:

$ 
P=\left[{\begin{array}{ccc} 
0 & 1& 0\\ 
0 & 0& 1\\ 
1 & 0& 0\\ 
\end{array} } \right] \Rightarrow \left[{\begin{array}{c}  
1\\ 
2\\ 
0\\ 
\end{array}}\right] 
$ 

D.h. es wird nur der Index des Wertes in jeder Zeile gespeichert, der nicht Null ist.

Wenn man dies macht, kann man später für die Lösung einfach den Vektor `bc = b[P]` verwenden, um auch $b$ richtig zu pivotisieren.  

Hinweise: 
- Die Pivot-Zeile lässt sich am einfachsten mit `np.argmax` finden. Die Zeile mit dem höchsten Betrag können Sie mit `np.abs` finden.

- Tauschen zweier Elemente in einem NumPy Array: `v[[a, b]] = v[[b, a]]`

- Versuchen Sie, die interne Vektorisierung von NumPy im Eliminationsschritt zu nutzen. Verwenden Sie dazu die Slicing-Syntax, um die innerste Schleife zu ersetzen.

_Points:_ 15

<span style="color: blue;">
  <h2>LU-Zerlegung mit Pivotisierung</h2>
</span>

<span style="color: blue;">
Im folgenden Code wird die LU-Zerlegung einer Matrix A mit Pivotisierung implementiert. Die Matrix A wird in die Form A = LU zerlegt, wobei L und U die untere und obere Dreiecksmatrix sind. Zusätzlich wird ein Pivot-Vektor zurückgegeben, der die Zeilenvertauschung anzeigt. Die Berechnung erfolgt in-place und nutzt NumPy für effiziente Vektorisierung.
</span>

In [8]:
def lu_decomposition(A):
    """L und U Matrix werden innerhalb von A Matrix berechnet und der Pivot-Vektor wird ausgegeben"""
    
    n = A.shape[0]
    P = np.arange(n)

    for k in range(n):
        # Pivotisierung: Index der Zeile mit größtem Betrag in Spalte k finden
        pivot_index = np.argmax(np.abs(A[k:n, k])) + k
        if A[pivot_index, k] == 0:
            raise ValueError("Matrix ist singulaer.")

        # Zeilen tauschen in A
        if pivot_index != k:
            A[[k, pivot_index], :] = A[[pivot_index, k], :]
            P[[k, pivot_index]] = P[[pivot_index, k]]

        # Berechnung L unterhalb der Diagonale
        A[k+1:n, k] /= A[k, k]

        # Berechnung U obere Dreiecksmatrix inklusive Diagonale
        A[k+1:n, k+1:n] -= np.outer(A[k+1:n, k], A[k, k+1:n])

    return P #Pivot-Vektor

<!-- BEGIN QUESTION -->

**Aufgabe:** Schreiben Sie eine Funktion `custom_solve`, welche mit Hilfe von `lu_decomposition` ein lineares Gleichungssystem der Form $A*x=b$ löst.  

Hinweis: Versuchen Sie, die interne Vektorisierung von NumPy zu nutzen. Verwenden Sie dazu die Slicing-Syntax, um die innerste Schleife zu ersetzen.

_Points:_ 12

In [9]:
def custom_solve(A, b):
    """Löst ein lineares Gleichungssystem A * x = b und gib Vektor x aus"""
    
    #Pivot-Vektor und A (welches LU-Zerlegung enthaelt) werden berechnet
    P = lu_decomposition(A)
    
    #L und U werden aus A isoliert
    n = A.shape[0]
    L = np.tril(A, k=-1) + np.eye(n)  # untere Dreiecksmatrix mit Einsen auf der Diagonale
    U = np.triu(A)                   # obere Dreiecksmatrix inklusive Diagonale
    
    #Lc = b[P] wird geloest fuer c
    #hier wird upper_triangular=False verwendet um eine allgemeine Matrix zu loesen
    c = b_solve(L, b[P], upper_triangular=False)
    
    #Ux = c wird geloest fuer x
    x = b_solve(U, c)
    
    return x

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

**Aufgabe:** Schreiben Sie eine Testfunktion, die überprüft, ob die Lösung ihrer Implementierung des Gauß-Verfahrens richtig ist.  
Testen Sie Gleichungssysteme der Größe $16 \times 16$, $32\times 32$ und $64 \times 64$.

Sie können zum Vergleich die SciPy Funktion `linalg.solve` verwenden:

_Points:_ 6

In [10]:
import scipy.linalg as linalg

def test_custom_solve(n):    
    """Vergleicht Ergebnisse von custom_solve mit linalg.solve mittels assert-Bedingung"""
    
    rng = np.random.default_rng(seed=123)
    A = rng.random((n, n))
    b = rng.random(n)
    linalg_x = linalg.solve(A, b)    
    custom_x = custom_solve(A, b)
    
    #sicherstellen, dass Ergebnisse gleich sind, float-Unpraezision ignorieren
    assert np.allclose(linalg_x, custom_x, atol=1e-6), f"Ergebnisse weichen zu stark ab bei Dimension {n}"
    
    print("Test bestanden fuer Dimension", n)
    
test_custom_solve(16)
test_custom_solve(32)
test_custom_solve(64)

Test bestanden fuer Dimension 16
Test bestanden fuer Dimension 32
Test bestanden fuer Dimension 64


<!-- END QUESTION -->

**Aufgabe:** Wir wollen nun weiter die *fertigen* SciPy Funktionen betrachten. Mit den Funktionen `lu` und `lu_factor` aus dem Packet `scipy.linalg `  können Sie eine LU-Dekomposition durchführen [[1]](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu_factor.html#scipy.linalg.lu_factor) und [[2]](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu.html).  
Beide Funktionen haben unter anderem den Parameter `overwrite_a`, der `False` oder `True` sein kann. 

Evaluieren Sie die Performance (Laufzeit) der LU-Zerlegung einer 4096x4096 Matrix, wenn Sie jeweils die Funktionen `lu_factor` und `lu` mit `overwrite_a = True` oder `overwrite_a = False` verwenden.

In der Liste `minimums` sollen die minimalen Laufzeiten in folgender Reihenfolge stehen:
1. `lu_factor` mit `overwrite_a = False`
2. `lu_factor` mit `overwrite_a = True`
3. `lu` mit `overwrite_a = False`
4. `lu` mit `overwrite_a = True`

Hinweis: Da die Matrix A bei der Verwendung von `overwrite_a = True` verändert wird, kann es zu einer Vereinfachung der Berechnung kommen, wenn die Funktion mehrfach nacheinander ausgeführt wird. Aus diesem Grund muss die `timeit` Funktion [repeat](https://docs.python.org/3/library/timeit.html#timeit.repeat) mit `repeat = 2` und `number = 1`  verwendet werden. Die Matrix wird im `setup` Parameter erstellt (aus Zufallswerten mit dem vorgegebenen "random number generator").

_Points:_ 6

In [12]:
from scipy.linalg import lu_factor, lu
from timeit import repeat

n = 4096
number = 1
repetitions = 2
# alle kürzesten Ausführungszeiten werden in minimums gespeichert
rng = np.random.default_rng(seed=123)
repeat_lambda = lambda x: repeat(x, setup="A = rng.random((n, n))", repeat=repetitions, number=number, globals=globals())
minimums = [min(repeat_lambda(x)) for x in ["lu_factor(A, overwrite_a=False)", "lu_factor(A, overwrite_a=True)", "lu(A, overwrite_a=False)", "lu(A, overwrite_a=True)"]]
minimums

[1.138857582998753, 1.0394668500011903, 1.5543745329996455, 1.3891434069992101]

<span style="color: blue;">
Die Luafzeitenanalyse zeigt, dass lu_faktor leicht schneller ist als lu.
Jeh nach System können die Laufzeiten variieren.
</span>

**Aufgabe:** Untersuchen Sie den Speicherverbrauch der 4 Varianten mit memray.

Um den Einfluss der anderen Berechnungen zu minimieren, empfehlen wir hier, die Berechnungen in eine neue Datei zu schreiben und mit `!` in einer Code-Zelle memray's Command-Line Interface zu verwenden. 

Am besten tracen Sie nur die LU Funktion mit einem Context Manager!
Verwenden Sie jeweils folgende Dateinamen als `file_name` Parameter in `memray.Tracker`:

- lu_inplace.bin
- lu_notinplace.bin
- lu_factor_inplace.bin
- lu_factor_notinplace.bin

Beantworten Sie bitte die folgenden Fragen:
* Welche Version verwendet am meisten Speicher? 
* Bei welcher Version ist der Speicherverbrauch am konstantesten? 

_Points:_ 9

In [13]:
#.py files werden erstellt für die jeweiligen Funktionen
# auch mit der line-magic %file einzeln möglich
for file_name in ["lu_inplace.bin", "lu_notinplace.bin", "lu_factor_inplace.bin", "lu_factor_notinplace.bin"]:
    with open(file_name.replace(".bin", ".py"), "w") as file:        
        file.write(
f"""import numpy as np
from scipy.linalg import lu_factor, lu
import memray

n = 4096
rng = np.random.default_rng(seed=123)
A = rng.random((n, n))

with memray.Tracker("{file_name}", native_traces=True):
    lu{"_factor" if "_factor" in file_name else ""}(A, overwrite_a={"_inplace" in file_name})
""")

<span style="color: blue;">
memray muss installiert und importierbar sein, es wurde version 1.17.1 verwendet.
</span>

In [ ]:
pip install memray==1.17.1

Note: you may need to restart the kernel to use updated packages.


In [14]:
import memray, os

#alle .py files werden ausgefuehrt in python3 um .bin Dateien zu erstellen
#diese enthalten die Infos zum Speicherbedarf
#falls sie schon im Verzeichnis existieren, müssen sie zuvor gelöscht werden
for file_name in ["lu_inplace.bin", "lu_notinplace.bin", "lu_factor_inplace.bin", "lu_factor_notinplace.bin"]:
    os.system(f"python3 {file_name.replace(".bin", ".py")}")

In [15]:
#Speichernutzung wird aus allen .bin files gelesen und hier zusammengefasst
for file_name in ["lu_inplace.bin", "lu_notinplace.bin", "lu_factor_inplace.bin", "lu_factor_notinplace.bin"]:
    reader = memray.FileReader(file_name)
    mem_use = 0

    for allocation in reader.get_allocation_records():
        mem_use += allocation.size

    print(f"Speichernutzung von {file_name[:-4]} : {round(mem_use / 2**20)}MB") 

Speichernutzung von lu_inplace : 432MB
Speichernutzung von lu_notinplace : 560MB
Speichernutzung von lu_factor_inplace : 176MB
Speichernutzung von lu_factor_notinplace : 176MB


<span style="color: blue;">
Die Funktionen mit Version lu_factor verwenden konstant nur 176MB an Speicher.

Funktionen mit Version lu verwenden deutlich mehr Speicher mit 432MB "inplace" und sogar 560MB "notinplace".

Der Speicherbedarf kann auf unterschiedlichen Systemen unterschiedlich ausfallen.
<span>


In [16]:
#Auch die command-line Funktion von memray kann verwendet werden
# äquivalent zu !memray stats lu_inplace.bin
os.system("memray stats lu_inplace.bin")

  Computing statistics... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00--:--
📏 Total allocations:
	16

📦 Total memory allocated:
	432.136MB

📊 Histogram of allocation size:
	min: 32.000B
	----------------------------------------
	< 147.000B : 2 ▇▇▇▇▇▇▇▇▇▇
	< 675.000B : 0 
	< 3.031KB  : 3 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 13.929KB : 1 ▇▇▇▇▇
	< 63.999KB : 5 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 294.066KB: 0 
	< 1.320MB  : 0 
	< 6.063MB  : 0 
	< 27.858MB : 1 ▇▇▇▇▇
	<=128.000MB: 4 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	----------------------------------------
	max: 128.000MB

📂 Allocator type distribution:
	 MALLOC: 13
	 CALLOC: 2
	 MMAP: 1

🥇 Top 5 largest allocating locations (by size):
	- lu:/home/codespace/.local/lib/python3.12/site-packages/scipy/linalg/_decomp_lu.py:355 -> 160.016MB
	- lu:/home/codespace/.local/lib/python3.12/site-packages/scipy/linalg/_decomp_lu.py:385 -> 128.000MB
	- lu:/home/codespace/.local/lib/python3.12/site-packages/scipy/linalg/_decomp_lu.py:354 -> 128.000MB
	- asarray_chkfinite:/home/codesp

0

In [17]:
#command-line Befehl zum Löschen aller erstellten files
os.system("rm -rf lu_*")

0